src/meta/src/barrier/worker.rs

```diff
@@ -311,19 +313,83 @@ impl<C: GlobalBarrierWorkerContext> GlobalBarrierWorker<C> {
+                        match event {
+                            CheckpointControlEvent::EnteringInitializing(entering_initializing) => {
+                                let database_id = entering_initializing.database_id();
+                                let error = merge_node_rpc_errors(&format!("database {} reset", database_id), entering_initializing.action.0.iter().filter_map(|(worker_id, resp)| {
+                                    resp.root_err.as_ref().map(|root_err| {
+                                        (*worker_id, ScoredError {
+                                            error: Status::internal(&root_err.err_msg),
+                                            score: Score(root_err.score)
+                                        })
+                                    })
+                                }));
+                                Self::report_collect_failure(&self.env, &error);
+                                if let Some(runtime_info) = self.context.reload_database_runtime_info(database_id).await? {
+                                    runtime_info.validate(database_id, &self.active_streaming_nodes).inspect_err(|e| {
+                                        warn!(database_id = database_id.database_id, err = ?e.as_report(), ?runtime_info, "reloaded database runtime info failed to validate");
+                                    })?;
+                                    entering_initializing.enter(runtime_info, &mut self.control_stream_manager);
+                                } else {
+                                    info!(database_id = database_id.database_id, "database removed after reloading empty runtime info");
+                                    entering_initializing.remove();
+                                }
+                            }
+                            CheckpointControlEvent::EnteringRunning(entering_running) => {
+                                self.context.mark_ready(Some(entering_running.database_id()));
+                                entering_running.enter();
+                            }
+                        }
+                    };
+                    if let Err(e) = result {
+                        self.failure_recovery(e).await;
+                    }
+                }
+                (worker_id, resp_result) = self.control_stream_manager.next_response() => {
+                     let resp = match resp_result {
+                        Err(e) => {
                             if self.checkpoint_control.is_failed_at_worker_err(worker_id) {
                                 let errors = self.control_stream_manager.collect_errors(worker_id, e).await;
                                 let err = merge_node_rpc_errors("get error from control stream", errors);
-                                self.report_collect_failure(&err);
+                                Self::report_collect_failure(&self.env, &err);
                                 self.failure_recovery(err).await;
```

Comment:
Global recovery may still be triggered in cases such as manually triggering recovery, and some internal errors of global barrier manager such as failed to commit epoch.

Reply:
I am not saying that we no longer have global recovery. My point is it is better that the global recovery is done by triggering DB recovery for each existing DB.

Reply:
Discussed offline. In future PRs, the logic when handling bidi-stream failure will become something similar to:
> - Get the databases assigned to the failing CN
> - Trigger database recovery for each of those databases


The reason why this PR still triggers a global recovery is because the CN side per database recovery logic is not included in this PR so to avoid breaking CI, we need to keep the logic global recovery separately in this PR.